services/web/server/src/simcore_service_webserver/snapshots_api_handlers.py

```diff
@@ -0,0 +1,255 @@
+            logger.debug(err, exc_info=True)
+            raise web.HTTPUnprocessableEntity(
+                text=json_dumps({"error": err.errors()}),
+                content_type="application/json",
+            ) from err
+
+        except ProjectNotFoundError as err:
+            logger.debug(err, exc_info=True)
+            raise web.HTTPNotFound(
+                reason=f"Project not found {err.project_uuid} or not accessible. Skipping snapshot"
+            ) from err
+
+    return wrapped
+
+
+def create_url_for_function(request: web.Request) -> Callable:
+    app = request.app
+
+    def url_for(router_name: str, **params) -> Optional[str]:
+        try:
+            rel_url: URL = app.router[router_name].url_for(
+                **{k: str(v) for k, v in params.items()}
+            )
+            url = request.url.origin().with_path(str(rel_url))
+            return str(url)
+        except KeyError:
+            return None
+
+    return url_for
+
+
+# FIXME: access rights using same approach as in access_layer.py in storage.
+# A user can only check snapshots (subresource) of its project (parent resource)
+
+
+# API ROUTES HANDLERS ---------------------------------------------------------
+routes = web.RouteTableDef()
+
+
+@routes.get(
+    f"/{vtag}/projects/{{project_id}}/snapshots",
+    name="list_project_snapshots_handler",
+)
+@login_required
+@permission_required("project.read")
+@handle_request_errors
+async def list_project_snapshots_handler(request: web.Request):
+    """
+    Lists references on project snapshots
+    """
+    snapshots_repo = SnapshotsRepository(request)
+    url_for = create_url_for_function(request)
+
+    @validate_arguments
+    async def _list_snapshots(project_id: UUID) -> List[Snapshot]:
+        # project_id is param-project?
+        # TODO: add pagination
+        # TODO: optimizaiton will grow snapshots of a project with time!
+        #
+        snapshots_orm = await snapshots_repo.list(project_id)
+        # snapshots:
+        #   - ordered (iterations!)
+        #   - have a parent project with all the parametrization
+
+        return [Snapshot.from_orm(obj) for obj in snapshots_orm]
+
+    snapshots: List[Snapshot] = await _list_snapshots(
+        project_id=request.match_info["project_id"],  # type: ignore
+    )
+    # TODO: async for snapshot in await list_snapshot is the same?
+
+    data = [SnapshotItem.from_snapshot(snp, url_for) for snp in snapshots]
+    return enveloped_response(data)
+
+
+@routes.get(
+    f"/{vtag}/projects/{{project_id}}/snapshots/{{snapshot_id}}",
+    name="get_project_snapshot_handler",
+)
+@login_required
+@permission_required("project.read")
+@handle_request_errors
+async def get_project_snapshot_handler(request: web.Request):
+    snapshots_repo = SnapshotsRepository(request)
+    url_for = create_url_for_function(request)
+
+    @validate_arguments
+    async def _get_snapshot(project_id: UUID, snapshot_id: str) -> Snapshot:
```

Comment:
ok I see the point. I'm not sure it is so much better now. it takes a lot of space. this is there to replace the extract_and_validate method? maybe for the next iteration we could in that direction? or maybe fuse the best of both worlds...

Reply:
``extract_and_validate`` will eventually *be removed* when the ``openapi-core`` library goes away. Here the idea is using pydantic validation + aiohttp request members to validate the request.